File: processor.cpp
#include <iostream>
#include <string>
#include <vector>
#include <cstdint>
#include "./core.cpp"


class processor{
    public:
    int clock =0;
    int clock_core1 = 0;
    int clock_core2 = 0;
    Core cores[2] = {Core(),Core()};
    int8_t memory1[2048] = {0};
    int8_t memory2[2048] = {0};
    int t=1,t1=1;
    
    void run()
    {   
        while(t1|t!=0)
        {
            t=cores[0].RunPipeline(memory1);
            t1=cores[1].RunPipeline(memory2);
            clock++;
            if(t == 0 && clock_core1 == 0)
            {
                clock_core1 = clock;
            }
            if(t1 == 0 && clock_core2 == 0)
            {
                clock_core2 = clock;
            }
        }
    }
    void printMemory()
    {
        int count = 0;
        std::cout << "First 2048 bytes" << std::endl;
        std::cout << "------------------------------------------------------------------------------------------------------"<<std::endl;
        for(int i = 0; i < 2048; i++)
        {
            std::cout << "|" << (int)memory1[i] << " ";
            count++;
            if((i+1) % 32 == 0)
            {
                std::cout << "|" << std::endl;
            }
        }
        std::cout << "------------------------------------------------------------------------------------------------------" <<std::endl;
        std::cout << "Second 2048 bytes" << std::endl;
        std::cout << "------------------------------------------------------------------------------------------------------"<<std::endl;
        for(int i = 0; i < 2048; i++)
        {
            std::cout << "|" << (int)memory2[i] << " ";
            count++;
            if((i+1) % 32 == 0)
            {
                std::cout << "|" << std::endl;
            }
        }
        std::cout << "------------------------------------------------------------------------------------------------------" <<std::endl;
    }
};

File: Processor.py
import array
import numpy as np
from Core import Core
from Cache import Cache
import pickle


class Processor:

    
    def __init__(self,cacheSize,blockSize,associativity,accesslatencyofCache,ReplacementPolicy):
        self.cache = Cache(cacheSize=cacheSize,blockSize=blockSize,associativity=associativity,ReplacementPolicy=ReplacementPolicy)
        self.accesslatencyofCache = accesslatencyofCache
        self.memory1 = np.zeros(2048, dtype=np.uint8)
        self.memory2 = np.zeros(2048, dtype=np.uint8)
        self.clock = 0
        self.clock1 = 0
        self.clock2 = 0
        self.Core1 = Core()
        self.Core2 = Core()
        with open("processor_state.txt", "w") as file:
            file.close()
        # print("Cache got Created")


        

    def run(self,latencies,end_pc1,end_pc2,dataForward):
        t,t1=True,True
        pr1,pr2 = False,False
        while t1 or t:
            print(self.clock)
            with open("processor_state.txt", "a") as file:
              file.write(f"cls.clock: {self.clock}\n")
              file.close()
            t = self.Core1.run(self.memory1,latencies,end_pc1,dataForward,cache=self.cache,memOffset=0,accesslatencyofCache=self.accesslatencyofCache)
            t1 = self.Core2.run(self.memory2,latencies,end_pc2,dataForward,cache=self.cache,memOffset=2048,accesslatencyofCache=self.accesslatencyofCache)
            if t==0 and not pr1:
                self.clock1=self.clock
                pr1 = True
            if t1==0 and not pr2:
                self.clock2=self.clock
                pr2 = True
            self.clock+=1


    def run1(self,latencies,end_pc1,end_pc2):
        # t,t1=True,False
        # while t1 or t:
        print(self.clock)
        t = self.Core1.run(self.memory1,latencies,end_pc1)
        #t1 = cls.Core2.run(cls.memory2,latencies,end_pc2)
        if t==0:
            self.clock1=self.clock
        # if t1==0:
        #     cls.clock2=cls.clock
        self.clock+=1


    
